Docs/default theme page

hbuchel opened the pull request:
<!--
Please make sure to read the Pull Request Guidelines:
https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md
-->

#### Description of changes

This PR adds a new Default theme page under Theming which will allow users to view some* of our default theme tokens. The new Default theme page has some links for exploring the tokens as well as some more information about how to reference tokens and our token structure.

Preview: https://docs-default-theme-page.d1scyzr8r1b4fk.amplifyapp.com/react/theming/default-theme

*Currently we have some for color, sizes, and typography.

###  New components:
**[TokenList](https://github.com/aws-amplify/amplify-ui/pull/2023/files#diff-a63282ca87887deaea5e5e775f099ade39721697286f1303c57d88a058fb7584)**: outputs a list of tokens with some sort of visual element depending on the namespace passed (color, space, etc).

**Usage**
```
<TokenList namespace="colors" childNamespace="neutral" />
```

<img width="680" alt="Screen Shot 2022-06-07 at 2 52 59 PM" src="https://user-images.githubusercontent.com/376920/172460181-cf8a1e72-6b76-4fb6-bb89-92251f212cbc.png">


**[CardLink](https://github.com/aws-amplify/amplify-ui/pull/2023/files#diff-8625a2493531b1fc0a82d7a0a07f1f03f0b225fabc0f6a220bf84b3ca7ad947a)**: Outputs a fancy clickable card link we can use throughout the docs to call attention to other pages. 

**Usage**
```
<CardLink 
  img={<MdOutlineAutoAwesome />}
  href="/react/theming#getting-started" 
  title="Theming overview" 
  desc="Learn more about theming in our Getting Started guide" />
```
<img width="975" alt="Screen Shot 2022-06-07 at 2 55 13 PM" src="https://user-images.githubusercontent.com/376920/172460582-d61fb3fc-4501-41d8-adc0-9f02af9ddc30.png">

Works well in a grid.

<img width="1105" alt="Screen Shot 2022-06-07 at 2 56 38 PM" src="https://user-images.githubusercontent.com/376920/172460790-2c0ee793-5209-43ed-80a3-91dbf50e4a81.png">

#### Issue #, if available


#### Not Included

- Sidebar needs style for tertiary pages so they look nested (already tracked in Asana).

#### Description of how you validated changes

#### Checklist

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are updated
- [x] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated
- [x] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

hbuchel commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,135 @@
+  ColorBlock,
+} from './TokenBlocks';
+
+export function createTokenList(tokens) {
+  // Creates a flattened array out of the token object passed to createTokenList()
+  let tokenList = [];
+  function iterateGroup(group) {
+    Object.values(group).forEach((value) => {
+      if (typeof value === 'object' && value.hasOwnProperty('name')) {
+        tokenList.push({
+          ...value,
+        });
+      } else {
+        iterateGroup(value);
+      }
+    });
+  }
+  if (tokens.hasOwnProperty('name')) {
+    tokenList.push(tokens);
+  } else {
+    iterateGroup(tokens);
+  }
+  return tokenList;
+}
```
Any feedback on a better way to do this? We need a nice flat array to map over for TokenList, and `tokens` can be different levels of nested-ness (e.g. tokens.space vs tokens.colors.neutral)

zchenwei replied:
I think this is good

hbuchel commented on docs/src/pages/[platform]/theming/default-theme/sizes/web.mdx:
````diff
@@ -0,0 +1,22 @@
+// Example usage
+{ tokens.radii.small }
+```
+<TokenList namespace="radii" />
````
Radii list is currently missing it's visual element.

reesscot commented on docs/src/pages/[platform]/theming/default-theme/typography/web.mdx:
```diff
@@ -0,0 +1,97 @@
+<TokenList namespace="lineHeights" />
+
+## Font families
+<TokenList namespace="fonts" />
```
Wondering if we should include a basic explanation of the difference between static and variable fonts? Link to MDN?

hbuchel replied:
Good idea, added a note to explain what they are with a link to MDN.

hbuchel commented:
@reesscot thanks! some updates below.

> 
> 1. Hard to read the Button in dark mode (may need to pass default dark mode override?)

I updated this to pass it a text color so it's not as weird in dark mode with our nested ThemeProvider.

> 2. Something is going on with syntax highlighting:

Fixed! Accidentally set the syntax to css.

> 3. `<sup>` element should be used for *

Thanks, fixed in the latest! (probably still building at the time of this comment on the preview link)

calebpollman reviewed (commented):
@hbuchel This looks so good! Very impressive 😄 

Before diving in for a thorough review, have a couple questions:
- Have we considered how this feature will scale to document non-web platform themes (React Native, iOS, Android) in the future?
- Not seeing any typing for the new functions or components, is that intentional?

zchenwei commented on docs/src/styles/docs/tokenList.scss:
```diff
@@ -0,0 +1,151 @@
+.docs-colorBlock {
+  width: var(--swatch-size);
+  height: var(--swatch-size);
+  border: 1px solid rgba(0,0,0,.1);
```
Can we use `var(--amplify-colors-overlay-xx)` for border color? Or maybe adjust it to 20. That way the border color can be flipped in dark mode as well. It is not friendly to some color in dark mode now.

![Screen Shot 2022-06-09 at 3 18 16 PM](https://user-images.githubusercontent.com/40295569/172954704-507d67b1-a39a-45cb-bdb7-3f32270c8b05.png)


With `var(--amplify-colors-overlay-20)`

![Screen Shot 2022-06-09 at 3 17 52 PM](https://user-images.githubusercontent.com/40295569/172954664-6761d1b1-808b-469e-86e2-20f03b10bba7.png)

zchenwei commented on docs/src/pages/[platform]/theming/default-theme/icons/Colors.tsx:
```diff
@@ -0,0 +1,26 @@
+
+export function Colors() {
+  return (
+    <Icon ariaLabel="" viewBox={{ width: 128, height: 128 }}>
```
Oh that's awesome. I was thinking you get the icons somewhere else but you built them by yourself

zchenwei commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,166 @@
+  let tokenList = [];
+  function iterateGroup(group) {
+    Object.values(group).forEach((value) => {
+      if (typeof value === 'object' && value.hasOwnProperty('name')) {
```
You can use this util `import { isDesignToken } from '@aws-amplify/ui';` to replace the conditional

hbuchel replied:
Oh perfect! Added that to both places.

zchenwei commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,166 @@
+      }
+    });
+  }
+  if (tokens.hasOwnProperty('name')) {
```
```suggestion
  if (isDesignToken(tokens)) {
```

zchenwei reviewed (commented):
Thanks for adding this nice page! Just a few minor comments

calebpollman commented on docs/src/pages/[platform]/theming/default-theme/icons/Colors.tsx:
```diff
@@ -0,0 +1,26 @@
+import { Icon } from '@aws-amplify/ui-react';
+
+export function Colors() {
```
Might be worth giving this a less generic name since this is a very broad namespace, for example `ColorsThemeIcon` (admittedly not a great name 😆 ). Same with on `Sizes` and `Typography`

hbuchel replied:
Yeah agreed, updated in latest.

calebpollman commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+  | 'fontSizes'
+  | 'fontWeights';
+
+export function createTokenList(tokens) {
```
Suggest moving in to a _components/Theming/utils.ts_ file, and adding strict-ish typing to make reading (and future refactoring if needed) easier

hbuchel replied:
I ended up moving this to a utils file, though typing could use some work (see [this comment](https://github.com/aws-amplify/amplify-ui/pull/2023#discussion_r895903948) here)

calebpollman commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+  }
+
+  // Create a more iteratable list (array) from our tokens
+  const tokenList = createTokenList(tokenNamespace);
```
It sort of feels like the above logic should be handled in `createTokenList`, but maybe not. Thoughts?

hbuchel replied:
Hmm I'm not sure the best way to handle this one because we need access to `tokens` from the useTheme() hook in order to determine what object to squash into a tokenList. I think this is something we can iterate on.

calebpollman commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+export function TokenItem({ variation, children }: TokenItemProps) {
+  return (
+    <li>
+      <div className={`docs-tokenItem docs-tokenItem--${variation}`}>
```
Would it make sense to just move the `className` to the `li` and remove this `div`? Just asking out of curiosity

hbuchel replied:
Yeah that makes sense, we don't need the extra div. Updated  in latest.

calebpollman reviewed (commented):
Great work overall! Left some questions/comments/nits on the React code but don't think any of them are particularly blocking.

calebpollman commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+
+  return (
+    <ul className="docs-tokenGroup docs-tokenGroup--list">
+      {tokenList.map((tokenItem) => {
```
Or just could 😄

hbuchel replied:
Updated this in latest to use useMemo for the tokenList mapping.

calebpollman commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+  RadiusBlock,
+} from './TokenBlocks';
+
+type Namespaces =
```
Just a follow up question, is it possible to derive this directly from the `Theme`? Just thinking that it would help give us some level of type safety/prevent deviation if a change is made in the `Theme` in the future

hbuchel replied:
I think I might leave cleaning up the Types further until post marketing launch. I was able to pull in some types directly from the @aws-amplify/ui package (like ColorBlock uses ColorValue, LineHeightBlock uses LineHeightValue) but ran into some missing exports when I tried to do this for others like [Tokens](https://github.com/aws-amplify/amplify-ui/blob/main/packages/ui/src/theme/tokens/index.ts#L25). I think we can find some ways around this though  (especially as we'd like to expand on showing component specific tokens at some point and having stronger typing will be nice for that).

calebpollman commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+  return <div className="docs-tokenItem-meta">{children}</div>;
+}
+
+type TokenListProps = {
+  namespace: Namespaces;
+  // TODO: better type for childNamespace? This should be children
+  // of whatever namespace you chose. e.g. namespace: 'colors', childNamespace: 'brand,primary'
+  childNamespace?: string;
+};
```
Had a follow up thought on this, there might be value in adding a utility type union for the `childNamespace` key, for example:
```
// would be best derived from `Theme` if possible
type ChildNamespace = 'primary' | 'secondary` | 'background' | 'etc';
```

We could then type `TokenListProps` as:
```type TokenListProps = {
  namespace: Namespaces;
  childNamespaces?: ChildNamespace[];
};
```

Passing an array as `childNamespaces` would remove the need to call `trim` and `split` on line 114, and `ChildNamespace` would enforce stricter usage in the future

calebpollman commented on docs/src/pages/[platform]/theming/default-theme/colors/web.mdx:
```diff
@@ -0,0 +1,94 @@
+## Semantic
+
+### Font
+<TokenList namespace="colors" childNamespace="font" />
```
Since _.mdx_ files doesn't have TS support, was thinking that it might be helpful to make some utility components to use in this file, example:

```
// in TokensList.tsx
// contrived name :)
const TokensColorsFont = () => <TokenList namespace="colors" childNamespace="font" />;

// usage here
<TokensColorsFont />
```

What do you think?

hbuchel replied:
Yeah, I like that. Makes it easier to try out the types (especially if we expand these). I updated that in the latest.

hbuchel commented on docs/src/components/Theming/TokenBlocks.tsx:
```diff
@@ -0,0 +1,153 @@
+  return (
+    <div className="docs-radiusBlock">
+      <svg
+        width={`max(${minBlockSize}, calc(2 * ${value} + 1px))`}
```
Yeah that works great! Updated in latest, much easier to make sense of now in the svg code. thanks!

hbuchel commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+
+export function createTokenList(tokens) {
+  // Creates a flattened array out of the token object passed to createTokenList()
+  let tokenList = [];
```
Yeah, updated in latest.

hbuchel commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+      case 'radii':
+        return <RadiusBlock value={value} />;
+      default:
+        return <div></div>;
```
I'll leave this as a GenericBlock return (updated in latest), if/when we add more namespace options we'll have something to start with so the grid is not messed up.

hbuchel commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+   * Depending on the namespace,  render a different visual
+   * element for the token list (like a color swatch for 'colors')
+   */
+  function renderVisualElement(value) {
```
Yeah, I like that! Updated in latest.

hbuchel commented on docs/src/components/Theming/TokenList.tsx:
```diff
@@ -0,0 +1,169 @@
+        const { name, path, value } = tokenItem;
+
+        // Remove the first path fragment which is the same as the namespace.
+        path.shift();
```
I ended up moving this out of TokenList and into TokenPath (at least for now).

zchenwei reviewed (approved):
LGTM

joebuono commented on docs/src/pages/[platform]/theming/default-theme/web.mdx:
```diff
@@ -0,0 +1,89 @@
+import { ColorsIcon, TypographyIcon, SizesIcon } from './icons';
+import { MdOutlineAutoAwesome } from 'react-icons/md';
+
+Amplify UI follows the [System UI Theme Specification](https://system-ui.com/theme/) with some modifications. This allows for a consistent pattern when defining our default tokens for properties such as color, fontSizes, border radius, and more. In addition to the reference here, you can browse our [default tokens](https://github.com/aws-amplify/amplify-ui/tree/main/packages/ui/src/theme/tokens)  ...
```
The System UI link appears to be broken for me. 
nit: "fontSizes" => "font size"
nit: "Github" => "GitHub"

joebuono commented on docs/src/pages/[platform]/theming/default-theme/web.mdx:
````diff
@@ -0,0 +1,89 @@
+  }
+```
+
+All of our tokens follow this pattern: a top level namespace that is further defined by a scale, each which has their own unique value. In the case of our space design tokens, that scale is based on a size: small, medium, large, etc. Our [color tokens](/react/theming/default-theme/colors#neutral), on the other hand, are further defined by hue (red, blue, yellow, etc), and then by a numerical scal ...
````
Not necessarily related to this PR, but I've been curious as to why the numerical scale for colors does not included 30, 50 or 70 @dbanksdesign

joebuono commented on docs/src/pages/[platform]/theming/default-theme/web.mdx:
````diff
@@ -0,0 +1,89 @@
+};
+```
+
+The space token definition also gives us an idea of how the css custom properties associated with each token are named. Amplify UI uses StyleDictionary to generate the associated css custom properties for each token. For the small space token, this would be `--amplify-space-small` and we can use it in your component:
````
Add external link to Style Dictionary?

nit: "css" to "CSS"

joebuono reviewed (approved):
So beautiful! Just a few small nits.
